### PR TITLE
Add bank & multiport support in TypeScript code generator with > 30 multiport tests.

### PR DESCRIPTION
**Pull request by hokeun**
Relevant PR: https://github.com/lf-lang/reactor-ts/pull/81
Relevant Issue: https://github.com/lf-lang/reactor-ts/issues/57

### CODE REVIEW

**Review by edwardalee — COMMENTED** [review]
Looks like a good start to me! I notice that code handling connections has no provision for the `interleaved` keyword. This will become an issue when banks are supported.

One high level question: This implementation has quite a bit of `if (p.isMultiport()) {...} else {...}` in the Kotlin code generator.  Would it make more sense to be doing these checks in reactor-ts? It seems that this realization is putting much of the work into the code generator, as done by CGenerator, rather than in the runtime. Will this limit the ability to support mutations?

I did not review the changes to reactor-ts, but some look suspicious. Is package-lock.json really supposed to be in the repo? It was removed from .gitignore.  What are the `__tests__` directories?

Please take all this with a grain of salt because I'm quite unfamiliar with this code.

File: org.lflang/src/org/lflang/AstExtensions.kt [hunk 1]
```diff
@@ -373,6 +373,11 @@ val Action.isPhysical get() = this.origin == ActionOrigin.PHYSICAL
  */
 val Port.isMultiport get() = JavaAstUtils.isMultiport(this)
 
+/**
+ * Return true if the receiving is a port and a multiport.
```
[the remark this answers is not in the document]
Comment (hokeun) [reply org.lflang/src/org/lflang/AstExtensions.kt:377]: Thanks for catching!

**Comment by hokeun** [comment]
> Looks like a good start to me!

First of all, thank you for the great comments!

> I notice that code handling connections has no provision for the `interleaved` keyword. This will become an issue when banks are supported.

I understand that the `interleaved` keyword should be taken care of. I believe it shouldn't be too difficult to support that with the current design using an Array of in/out ports for multiports. (I can imagine we just use Arrays of reactors for banks and switch how we order the ports/banks.) I also took note in the issue [here](https://github.com/lf-lang/reactor-ts/issues/57#issuecomment-1035599271).

> One high level question: This implementation has quite a bit of `if (p.isMultiport()) {...} else {...}` in the Kotlin code generator. Would it make more sense to be doing these checks in reactor-ts? It seems that this realization is putting much of the work into the code generator, as done by CGenerator, rather than in the runtime. Will this limit the ability to support mutations?

I think I can move some of the generated code to the reactor-ts runtime (maybe as functions or classes), however, some checks might have to stay in the code generator. Let me talk to Marten and try to move those codes to the runtime as much as I can. Thanks!

> I did not review the changes to reactor-ts, but some look suspicious. Is package-lock.json really supposed to be in the repo? It was removed from .gitignore. What are the `__tests__` directories?
> 
> Please take all this with a grain of salt because I'm quite unfamiliar with this code.

I agree that `package-lock.json` should probably not be in the repo. I think the `__tests__` directories are for unit tests of the reactor-ts runtime. I'm not fully aware of why those changes were made since I was just taking the recent commit of the reactor-ts runtime, but I'll try to clean them up.

**Comment by lhstrh** [comment]
https://docs.npmjs.com/cli/v6/configuring-npm/package-locks/#using-locked-packages

**Comment by hokeun** [comment]
> One high level question: This implementation has quite a bit of `if (p.isMultiport()) {...} else {...}` in the Kotlin code generator. Would it make more sense to be doing these checks in reactor-ts? It seems that this realization is putting much of the work into the code generator, as done by CGenerator, rather than in the runtime. Will this limit the ability to support mutations?

I tried to address this issue by moving the work from the code generator to the runtime in https://github.com/lf-lang/lingua-franca/pull/942/commits/8d5b52c89dc876e7b752128617d7c92d3d1b6db8.

**Comment by cmnrd** [comment]
Just for reference, the C++ generator uses the library method [`bind_multiple_ports`](https://github.com/lf-lang/lingua-franca/blob/442cc60b9bb7733f925b13527d60364246c941ec/org.lflang/src/lib/cpp/lfutil.hh#L64), to create multiport connections. With this method, all the code generator needs to do, is to assemble a vector of all ports listed on the left and all ports listed on the right of the connection operator.

After a quick glance over the code, I wonder: does it support listing multiple ports on the connection iterator? I mean something like this: `a, b.out, c.out -> d.in, e.in, f`

**Comment by hokeun** [comment]
> Just for reference, the C++ generator uses the library method [`bind_multiple_ports`](https://github.com/lf-lang/lingua-franca/blob/442cc60b9bb7733f925b13527d60364246c941ec/org.lflang/src/lib/cpp/lfutil.hh#L64), to create multiport connections. With this method, all the code generator needs to do, is to assemble a vector of all ports listed on the left and all ports listed on the right of the connection operator.

Thanks for the comment, @cmnrd ! Yes, I checked `bind_multiple_ports` for TS multiport implementation - [here](https://github.com/lf-lang/reactor-ts/issues/57#issuecomment-1030715978) you can see how it was considered. 

> After a quick glance over the code, I wonder: does it support listing multiple ports on the connection iterator? I mean something like this: `a, b.out, c.out -> d.in, e.in, f`

I'm not sure what you meant by "connection iterator"; however, the current implementation (which is not complete) partially supports what you described (but not entirely) like, `b.out, c.out -> d.in, e.in` where b/c/d/e can be banks and out/in can be multiports. We surely would want to support all the cases in the multiport tests. I'd appreciate it if you have any insights or suggestions from your experience in implementing multiports/banks in cpp. :)

**Comment by cmnrd** [comment]
> I'm not sure what you meant by "connection iterator"; however, the current implementation (which is not complete) partially supports what you described (but not entirely) like, b.out, c.out -> d.in, e.in where b/c/d/e can be banks and out/in can be multiports. We surely would want to support all the cases in the multiport tests. I'd appreciate it if you have any insights or suggestions from your experience in implementing multiports/banks in cpp. :)

I actually meant connection *operator* ;)

In C++ we replicate the reactor hierarchy including all ports. This means that for each port in LF, there is also a corresponding port in the generated C++ code. This is unlike the C generator, which might omit any ports in the middle of a connection that are not used as a trigger. Whether you can re-use the C++ solution, depends on whether you also replicate all the ports 1-to-1 in TS or you optimize some ports away as  is done in C.

Roughly, `b.out, c.out -> d.in, e.in` would be translated to C++ code like this:
```
std::vector<std::port<...>> left_ports;
std::vector<std::port<...>> right_ports;
\\ add all ports in b.out to left_ports
\\ add all ports in c.out to left_ports
\\ add all ports in d.in to right_ports
\\ add all ports in e.in to right_ports
lfutil::bind_multiple_ports(left_ports, right_ports, false)
```
Depending on the nature of the listed ports, `\\ add all ports` will do different things. If `b.out` is a single port, the generator will simply produce `left_ports.push_back(&b.out)`. If it is a multiport, it will produce a for loop iterating over all port instances, adding them individually. If it is a port with in a bank, it will also produce a for loop, but iterate over the bank instances. If it is a multiport within a bank two nested for loops will be generated. This latter case can be modified by the `interlvead` keyword which will switch the inner and outer loops.

This mechanism allows the code generator to cover all the possible cases and list all ports in the proper order in a single vector. The same happens for the right side. Finally, the call to `bind_multiple_ports` establishes all the connections. In case the `(...)+` operator was used, the boolean argument to `bind_multiple_ports` will be set to true and the connection algorithm implemented by the function adjusts accordingly.

All in all, this strategy enables a relatively simple implementation of all the complicated combinations that could appear n the connection operator. The code generator is responsible for providing a data structure that lists all ports in the relevant order, and the runtime method is then responsible for establishing the connections.

File: org.lflang/src/org/lflang/generator/ts/TSConnectionGenerator.kt [hunk 2]
```diff
@@ -14,30 +22,70 @@ class TSConnectionGenerator (
...
+                        "(__acc, __val) => __acc.concat(__val.${port.variable.name}), " +
+                        "new Array<${getPortTypeName(port.variable as Port, genericTypeMap)}>(0))"
+                return portName
+            }
+            portName += port.container.name + "."
+        }
+        portName += port.variable.name
+        return "this.$portName"
+    }
+
```
Comment (hokeun) [inline org.lflang/src/org/lflang/generator/ts/TSConnectionGenerator.kt:46]: Add comments.

**Review by lhstrh — APPROVED** [review]

**Review by lhstrh — APPROVED** [review]

**Pull request merged**